src/measure_theory/group/fundamental_domain.lean

```diff
@@ -0,0 +1,106 @@
+  [has_vadd G α] [measurable_space α] (s : set α) (μ : measure α . volume_tac) : Prop :=
+(measurable_set : measurable_set s)
+(ae_covers : ∀ᵐ x ∂μ, ∃ g : G, g +ᵥ x ∈ s)
+(ae_disjoint : ∀ g ≠ (0 : G), μ ((g +ᵥ s) ∩ s) = 0)
+
+/-- A measurable set `s` is a *fundamental domain* for an action of a group `G` on a measurable
+space `α` with respect to a measure `α` if the sets `g • s`, `g : G`, are pairwise a.e. disjoint and
+cover the whole space. -/
+@[protect_proj, to_additive is_add_fundamental_domain]
+structure is_fundamental_domain (G : Type*) {α : Type*} [has_one G] [has_scalar G α]
+  [measurable_space α] (s : set α) (μ : measure α . volume_tac) : Prop :=
+(measurable_set : measurable_set s)
+(ae_covers : ∀ᵐ x ∂μ, ∃ g : G, g • x ∈ s)
+(ae_disjoint : ∀ g ≠ (1 : G), μ ((g • s) ∩ s) = 0)
+
+namespace is_fundamental_domain
+
+variables {G α : Type*} [group G] [mul_action G α] [measurable_space α] {s t : set α}
+  {μ : measure α}
+
+@[to_additive] lemma Union_smul_ae_eq (h : is_fundamental_domain G s μ) :
+  (⋃ g : G, g • s) =ᵐ[μ] univ :=
+filter.eventually_eq_univ.2 $ h.ae_covers.mono $
+  λ x ⟨g, hg⟩, mem_Union.2 ⟨g⁻¹, _, hg, inv_smul_smul _ _⟩
+
+variables [measurable_space G] [has_measurable_smul G α]
+
+@[to_additive]
+lemma measurable_set_smul (h : is_fundamental_domain G s μ) (g : G) : measurable_set (g • s) :=
+h.measurable_set.const_smul g
+
+variables [smul_invariant_measure G α μ]
+
+@[to_additive] lemma pairwise_ae_disjoint (h : is_fundamental_domain G s μ) :
+  pairwise (λ g₁ g₂ : G, μ (g₁ • s ∩ g₂ • s) = 0) :=
+λ g₁ g₂ hne,
+calc μ (g₁ • s ∩ g₂ • s) = μ (g₂ • ((g₂⁻¹ * g₁) • s ∩ s)) :
+  by rw [smul_set_inter, ← mul_smul, mul_inv_cancel_left]
+... = μ ((g₂⁻¹ * g₁) • s ∩ s) : measure_smul_set _ _ _
+... = 0 : h.ae_disjoint _ $ mt inv_mul_eq_one.1 hne.symm
+
+variables [encodable G]
+
+@[to_additive] lemma measure_eq_tsum' (h : is_fundamental_domain G s μ) (t : set α) :
+  μ t = ∑' g : G, μ (t ∩ g • s) :=
+calc μ t = μ.restrict t (⋃ g : G, g • s) :
+  by rw [measure_congr (ae_mono measure.restrict_le_self h.Union_smul_ae_eq), restrict_apply_univ]
+... = ∑' g : G, μ.restrict t (g • s) :
+  measure_Union_of_null_inter h.measurable_set_smul $ h.pairwise_ae_disjoint.mono $
+    λ g₁ g₂ H, measure.restrict_le_self.absolutely_continuous H
+... = ∑' g : G, μ (t ∩ g • s) :
+  tsum_congr $ λ g, by rw [measure.restrict_apply (h.measurable_set_smul g), inter_comm]
+
+@[to_additive] lemma measure_eq_tsum (h : is_fundamental_domain G s μ) (t : set α) :
+  μ t = ∑' g : G, μ (g • t ∩ s) :=
+calc μ t = ∑' g : G, μ (t ∩ g • s) : h.measure_eq_tsum' t
+     ... = ∑' g : G, μ (t ∩ g⁻¹ • s) : ((equiv.inv G).tsum_eq _).symm
+     ... = ∑' g : G, μ (g • t ∩ s) :
+  tsum_congr $ λ g, by rw [← measure_smul_set g μ, smul_set_inter, smul_inv_smul]
+
+/-- If `s` and `t` are two fundamental domains of the same action, then their measures are equal. -/
+@[to_additive] protected lemma measure_congr (hs : is_fundamental_domain G s μ)
```

Comment:
This is not really a `congr` lemma. Maybe `measure_unique`?

Reply:
Renamed to `measure_eq`.